contracts/vanchor/src/contract.rs

```diff
@@ -0,0 +1,528 @@
+                output
+            };
+
+            // Compute hash of abi encoded ext_data, reduced into field from config
+            // Ensure that the passed external data hash matches the computed one
+            let mut ext_data_args = Vec::new();
+            let recipient_bytes = element_encoder(
+                &hex::decode(&ext_data.recipient).map_err(|_| ContractError::DecodeError)?,
+            );
+            let relayer_bytes = element_encoder(
+                &hex::decode(&ext_data.relayer).map_err(|_| ContractError::DecodeError)?,
+            );
+            let fee_bytes = element_encoder(&ext_data_fee.to_le_bytes());
+            let ext_amt_bytes = element_encoder(&ext_amt.to_le_bytes());
+            ext_data_args.extend_from_slice(&recipient_bytes);
+            ext_data_args.extend_from_slice(&relayer_bytes);
+            ext_data_args.extend_from_slice(&ext_amt_bytes);
+            ext_data_args.extend_from_slice(&fee_bytes);
+            ext_data_args.extend_from_slice(&ext_data.encrypted_output1);
+            ext_data_args.extend_from_slice(&ext_data.encrypted_output2);
+
+            let computed_ext_data_hash =
+                Keccak256::hash(&ext_data_args).map_err(|_| ContractError::HashError)?;
+            if computed_ext_data_hash != proof_data.ext_data_hash {
+                return Err(ContractError::Std(StdError::GenericErr {
+                    msg: "Invalid ext data".to_string(),
+                }));
+            }
+
+            let abs_ext_amt = ext_amt.unsigned_abs();
+            // Making sure that public amount and fee are correct
+            if Uint256::from(ext_data_fee) > vanchor.max_fee {
+                return Err(ContractError::Std(StdError::GenericErr {
+                    msg: "Invalid fee amount".to_string(),
+                }));
+            }
+
+            if Uint256::from(abs_ext_amt) > vanchor.max_ext_amt {
+                return Err(ContractError::Std(StdError::GenericErr {
+                    msg: "Invalid ext amount".to_string(),
+                }));
+            }
+
+            // Public amounnt can also be negative, in which
+            // case it would wrap around the field, so we should check if FIELD_SIZE -
+            // public_amount == proof_data.public_amount, in case of a negative ext_amount
+            let calc_public_amt = ext_amt - ext_data_fee as i128;
+            let calc_public_amt_bytes =
+                element_encoder(&ArkworksIntoFieldBn254::into_field(calc_public_amt));
+            if calc_public_amt_bytes != proof_data.public_amount {
+                return Err(ContractError::Std(StdError::GenericErr {
+                    msg: "Invalid public amount".to_string(),
+                }));
+            }
+
+            // Construct public inputs
+            let chain_id_type_bytes = element_encoder(
+                &compute_chain_id_type(vanchor.chain_id, &COSMOS_CHAIN_TYPE).to_le_bytes(),
```

Comment:
This should be BE bytes

Reply:
Hmm, I think it is "LE" bytes.
https://github.com/webb-tools/protocol-cosmwasm/blob/vanchor-implement/contracts/vanchor/src/test_util.rs#L201
This "test_util" is from "protocol-substrate" & it indicates that "LE" is used for chain id.